File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class TrieNode {
    public:
        bool isWord;
        TrieNode *child[26];
        TrieNode(){
            isWord = false;
            for( auto &a: child) a = nullptr;
        }
};

class Trie {
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode * p = root;
        for( auto &i : word){
            if(!p->child[i-'a']) p->child[i-'a'] = new TrieNode();
            p = p->child[i-'a'];
        }
        p->isWord = true;
    }
    
    bool search(string word, bool prefix = false) {
        TrieNode * p = root;
        for( auto &i : word){
            if(!p->child[i-'a']) return false;
            p = p->child[i-'a'];
        }
        if(prefix == false) return p->isWord;
        return true;
    }
    
    bool startsWith(string prefix) {
        return search(prefix, true);
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.isWord = False
        self.child = [None] * 26  # To store 26 lowercase English letters

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str) -> None:
        p = self.root
        for char in word:
            index = ord(char) - ord('a')  # Calculate index for the character
            if not p.child[index]:
                p.child[index] = TrieNode()
            p = p.child[index]
        p.isWord = True
    
    def search(self, word: str, prefix: bool = False) -> bool:
        p = self.root
        for char in word:
            index = ord(char) - ord('a')
            if not p.child[index]:
                return False
            p = p.child[index]
        if not prefix:
            return p.isWord
        return True
    
    def startsWith(self, prefix: str) -> bool:
        return self.search(prefix, True)


# Example usage:
# obj = Trie()
# obj.insert("word")
# param_2 = obj.search("word")
# param_3 = obj.startsWith("prefix")
